source/common/router/upstream_request.cc

```diff
@@ -125,7 +125,13 @@ UpstreamRequest::UpstreamRequest(RouterFilterInterface& parent,
       parent_.callbacks()->connection(), parent_.callbacks()->streamId(),
       parent_.callbacks()->account(), true, parent_.callbacks()->decoderBufferLimit(),
       *parent_.cluster(), *this);
-  parent_.cluster()->createFilterChain(*filter_manager_);
+  if (parent_.cluster()->filterChainLength() <= 1 && parent_.config().filterChainLength() > 0) {
```

Comment:
can you check out the comments I made on the internal version?  I don't see them applied here
/wait

Reply:
You're talking about having createFilterChain return bool, right? The thing is that clusters will always have the codec filter attached to them, and so `parent_.cluster()->createFilterChain()` would always return `true`. I guess something like 
```
if (parent_.cluster()->filterChainLength() > 1 // there is a nontrivial cluster filter chain
    || !parent_.config().createFilterChain(*filter_manager_) /* there is no router filter chain */ ) {
  parent_.cluster()->createFilterChain(*filter_manager_);
}
```
would do the same thing as what I've got above, but I don't know that it's clearer.

Another possibility is to just totally split the codec filter off of the filter chain; so the above would become:
```
parent_.cluster()->createFilterChain(...) || parent_.config().createFilterChain(...);
parent_.cluster()->appendCodecFilter(...);
```
The first line will only run the router filter chain creation if cluster filter chain creation returns false.

Reply:
oh yeah sorry I had it inverted and thought router trumped cluster so we could do something like we do for the upgrade filter chain but we can't because as you say the "nothing configured" case is empty.

OK new idea, which I'm writing in a hurry so poke me this afternoon if it doesn't make sense.

ow about adding a boolean to createFilterChain so we only create it if there's more than one entry, i.e if that filter chain is actually configured?  then in the long run we could

bool created = cluster->createUpgradeFilterChain(filter_manager)
if (!created) {
  created = cluster->createFilterChain(filter_manager_, true /*only_create_if_configured*/)
}
if (!created) { 
  created = router->createUpgradeFilterChain()
}
createFilterChain(filter_manager_, false /* only_create_if_configured */)

which I think is cleaner than a bunch of size checks?  alternately change your size check to isFilterChainConfigured but then we have to check existence of regular filter chains or best effort try-create upgrade filter chains, which I don't love.

Reply:
My other concern with this is that in the future we may add Filter Config to the UpstreamCodecFilter, in which case clusters could actually configure UpstreamCodecFilter. There wouldn't be a good way to distinguish between an UpstreamCodecFilter that has the default values, and one that has values assigned by the cluster.

Reply:
I actually see that as a feature and not a bug.  now that I think of it I really don't like length.
If you have a route config filterA filterB codec, and you want an override of "don't run any filters" how would you do it?   the override filter length is 1 so we'd skip?  I think we need to go by the user configuration proto - if one was present at all or if we used the default config, rather than the length of the filter chain, so again ideally the 'only if configured' bool but alternately a filterChainConfigured() method returning bool

Reply:
I think I prefer the `isCustomFilterChainConfigured()` method returning `bool`, personally.